src/runtime/task/mod.rs

```diff
@@ -265,3 +287,146 @@ impl Debug for TaskSet {
         write!(f, "}}")
     }
 }
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct VectorClock {
+    pub(crate) time: SmallVec<[u32; DEFAULT_INLINE_TASKS]>,
+}
+
+impl VectorClock {
+    pub(crate) fn new() -> Self {
+        Self { time: SmallVec::new() }
+    }
+
+    #[cfg(test)]
+    pub(crate) fn new_from(v: &[u32]) -> Self {
+        let time: SmallVec<[u32; DEFAULT_INLINE_TASKS]> = SmallVec::from(v);
+        Self { time }
+    }
+
+    // Zero extend clock to accommodate `task_id` tasks.
+    pub(crate) fn extend(&mut self, task_id: TaskId) {
+        let d = 1 + task_id.0 - self.time.len();
+        self.time.extend_from_slice(vec![0u32; d].as_slice());
+    }
+
+    pub(crate) fn increment(&mut self, task_id: TaskId) {
+        assert!(task_id.0 < self.time.len());
+        self.time[task_id.0] += 1;
+    }
+
+    // Update the clock of `self` with the clock from `other`
+    pub(crate) fn update(&mut self, other: &Self) {
+        let n1 = self.time.len();
+        let n2 = other.time.len();
+        for i in 0..std::cmp::min(n1, n2) {
+            self.time[i] = std::cmp::max(self.time[i], other.time[i])
+        }
+        for i in n1..n2 {
+            // could be empty
+            self.time.push(other.time[i]);
+        }
+    }
+
+    pub fn get(&self, i: usize) -> u32 {
+        assert!(i < self.time.len());
+        self.time[i]
+    }
+}
+
+impl<'a> IntoIterator for &'a VectorClock {
+    type Item = &'a u32;
+    type IntoIter = Iter<'a, u32>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.time.iter()
+    }
+}
+
+impl From<VectorClock> for SmallVec<[u32; DEFAULT_INLINE_TASKS]> {
+    fn from(v: VectorClock) -> SmallVec<[u32; DEFAULT_INLINE_TASKS]> {
+        v.time
+    }
+}
+
+impl<'a> From<&'a VectorClock> for &'a SmallVec<[u32; DEFAULT_INLINE_TASKS]> {
+    fn from(v: &'a VectorClock) -> &'a SmallVec<[u32; DEFAULT_INLINE_TASKS]> {
+        &v.time
+    }
+}
```

Comment:
I think these are all unused. Even if they weren't, `new_from` covers the need for `From` impls since `SmallVec<[T; _]>` derefs to `[T]`, and if we wanted the `IntoIterator` we should just instead implement `Deref<Target = [u32]>` or something.

Reply:
Yes, much nicer; fixed.